ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java

```diff
@@ -92,34 +100,42 @@
    * If content size is set in the header (i.e. not -1), the actual content size should match that value. Otherwise, an
    * exception will be thrown.
    * @param request the {@link HttpRequest} that needs to be wrapped.
+   * @param channel the {@link Channel} over which the {@code request} has been received.
    * @param nettyMetrics the {@link NettyMetrics} instance to use.
    * @throws IllegalArgumentException if {@code request} is null.
    * @throws RestServiceException if the {@link HttpMethod} defined in {@code request} is not recognized as a
    *                                {@link RestMethod}.
    */
-  public NettyRequest(HttpRequest request, NettyMetrics nettyMetrics)
+  public NettyRequest(HttpRequest request, Channel channel, NettyMetrics nettyMetrics)
       throws RestServiceException {
-    if (request == null) {
-      throw new IllegalArgumentException("Received null HttpRequest");
+    if (request == null || channel == null) {
+      throw new IllegalArgumentException("Received null argument(s)");
     }
     restRequestMetricsTracker.nioMetricsTracker.markRequestReceived();
+    this.request = request;
+    query = new QueryStringDecoder(request.getUri());
+    this.channel = channel;
+    savedMaxMessagesPerRead = channel.config().getMaxMessagesPerRead();
```

Comment:
Yes. Specifically this is the number of messages to read per `channel.read()` acc to the [documentation](http://netty.io/4.0/api/io/netty/channel/ChannelConfig.html#setMaxMessagesPerRead(int)). But each message should translate to a single `HttpContent`.